{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/zk/dmbrd2310jn8kdb96r649lr80000gn/T/ipykernel_875/2561380274.py:1: DeprecationWarning: \n",
      "Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),\n",
      "(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)\n",
      "but was not found to be installed on your system.\n",
      "If this would cause problems for you,\n",
      "please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466\n",
      "        \n",
      "  import pandas\n"
     ]
    }
   ],
   "source": [
    "import pandas \n",
    "import json\n",
    "\n",
    "def parse_five_prime_utr_variant_consequence(conseq_str):\n",
    "    \"\"\"\n",
    "    Parses the consequence str into a keyed dictionary as per\n",
    "    https://github.com/ImperialCardioGenetics/UTRannotator#the-detailed-annotation-for-each-consequence\n",
    "\n",
    "    \"\"\"\n",
    "    return {\n",
    "        annotation.split(':')[0]: annotation.split(':')[1]\n",
    "        for annotation in conseq_str.split(',')\n",
    "    }\n",
    "orfs = pandas.read_csv('../../../data/pipeline/ORFS_Features_1.0.tsv', sep=\"\\t\")\n",
    "orfs['ensembl_transcript_id'] = orfs['ensembl_transcript_id'].apply(\n",
    "        lambda x: x.split('.')[0]\n",
    "    )\n",
    "parsed_variants = pandas.read_csv('../../../data/pipeline/toy_dataset.txt', sep=\"\\t\")\n",
    "parsed_variants['variant_id'] = parsed_variants['#Uploaded_variation']\n",
    "\n",
    "sequences = pandas.read_csv('../../../data/pipeline/MANE_transcripts_v1.0.tsv', sep=\"\\t\")\n",
    "sequences['ensembl_transcript_id'] = sequences['transcript_id'].apply(lambda x: x.split('.')[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>five_prime_utr_length</th>\n",
       "      <th>three_prime_utr_length</th>\n",
       "      <th>num_five_prime_utr_exons</th>\n",
       "      <th>start_site_pos</th>\n",
       "      <th>cds_start</th>\n",
       "      <th>cds_end</th>\n",
       "      <th>cds_length</th>\n",
       "      <th>ensembl_transcript_id</th>\n",
       "      <th>annotations</th>\n",
       "      <th>...</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>description</th>\n",
       "      <th>build</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233.10</td>\n",
       "      <td>88</td>\n",
       "      <td>401</td>\n",
       "      <td>1</td>\n",
       "      <td>89</td>\n",
       "      <td>2</td>\n",
       "      <td>544</td>\n",
       "      <td>543</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>&gt;ENST00000000233.10 cdna chromosome:GRCh38:7:1...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000004059.11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>ADP ribosylation factor 5 [Source:HGNC Symbol;...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>7</td>\n",
       "      <td>127588411</td>\n",
       "      <td>127591700</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000412.8</td>\n",
       "      <td>159</td>\n",
       "      <td>1457</td>\n",
       "      <td>2</td>\n",
       "      <td>160</td>\n",
       "      <td>3</td>\n",
       "      <td>836</td>\n",
       "      <td>834</td>\n",
       "      <td>ENST00000000412</td>\n",
       "      <td>&gt;ENST00000000412.8 cdna chromosome:GRCh38:12:8...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000003056.8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>M6PR</td>\n",
       "      <td>mannose-6-phosphate receptor, cation dependent...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>12</td>\n",
       "      <td>8940361</td>\n",
       "      <td>8949645</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000442.11</td>\n",
       "      <td>225</td>\n",
       "      <td>777</td>\n",
       "      <td>2</td>\n",
       "      <td>226</td>\n",
       "      <td>3</td>\n",
       "      <td>1274</td>\n",
       "      <td>1272</td>\n",
       "      <td>ENST00000000442</td>\n",
       "      <td>&gt;ENST00000000442.11 cdna chromosome:GRCh38:11:...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000173153.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ESRRA</td>\n",
       "      <td>estrogen related receptor alpha [Source:HGNC S...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>11</td>\n",
       "      <td>64305524</td>\n",
       "      <td>64316743</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000001008.6</td>\n",
       "      <td>170</td>\n",
       "      <td>2165</td>\n",
       "      <td>1</td>\n",
       "      <td>171</td>\n",
       "      <td>2</td>\n",
       "      <td>1381</td>\n",
       "      <td>1380</td>\n",
       "      <td>ENST00000001008</td>\n",
       "      <td>&gt;ENST00000001008.6 cdna chromosome:GRCh38:12:2...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000004478.8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>FKBP prolyl isomerase 4 [Source:HGNC Symbol;Ac...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>12</td>\n",
       "      <td>2794970</td>\n",
       "      <td>2805423</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000001146.7</td>\n",
       "      <td>28</td>\n",
       "      <td>2989</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>2</td>\n",
       "      <td>1540</td>\n",
       "      <td>1539</td>\n",
       "      <td>ENST00000001146</td>\n",
       "      <td>&gt;ENST00000001146.7 cdna chromosome:GRCh38:2:72...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000003137.9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CYP26B1</td>\n",
       "      <td>cytochrome P450 family 26 subfamily B member 1...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>2</td>\n",
       "      <td>72129238</td>\n",
       "      <td>72147862</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19115</th>\n",
       "      <td>ENST00000699088.1</td>\n",
       "      <td>22</td>\n",
       "      <td>239</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>2</td>\n",
       "      <td>922</td>\n",
       "      <td>921</td>\n",
       "      <td>ENST00000699088</td>\n",
       "      <td>&gt;ENST00000699088.1 cdna chromosome:GRCh38:7:10...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000213413.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PVRIG</td>\n",
       "      <td>PVR related immunoglobulin domain containing [...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>7</td>\n",
       "      <td>100219949</td>\n",
       "      <td>100221490</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19116</th>\n",
       "      <td>ENST00000699294.1</td>\n",
       "      <td>487</td>\n",
       "      <td>2911</td>\n",
       "      <td>2</td>\n",
       "      <td>488</td>\n",
       "      <td>3</td>\n",
       "      <td>1676</td>\n",
       "      <td>1674</td>\n",
       "      <td>ENST00000699294</td>\n",
       "      <td>&gt;ENST00000699294.1 cdna chromosome:GRCh38:12:7...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000180881.21</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CAPS2</td>\n",
       "      <td>calcyphosine 2 [Source:HGNC Symbol;Acc:HGNC:16...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>12</td>\n",
       "      <td>75275979</td>\n",
       "      <td>75330324</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19117</th>\n",
       "      <td>ENST00000699333.1</td>\n",
       "      <td>181</td>\n",
       "      <td>5137</td>\n",
       "      <td>1</td>\n",
       "      <td>182</td>\n",
       "      <td>2</td>\n",
       "      <td>1270</td>\n",
       "      <td>1269</td>\n",
       "      <td>ENST00000699333</td>\n",
       "      <td>&gt;ENST00000699333.1 cdna chromosome:GRCh38:19:5...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000251369.10</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ZNF550</td>\n",
       "      <td>zinc finger protein 550 [Source:HGNC Symbol;Ac...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>19</td>\n",
       "      <td>57541838</td>\n",
       "      <td>57559863</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19118</th>\n",
       "      <td>ENST00000699524.1</td>\n",
       "      <td>170</td>\n",
       "      <td>1490</td>\n",
       "      <td>2</td>\n",
       "      <td>171</td>\n",
       "      <td>3</td>\n",
       "      <td>1325</td>\n",
       "      <td>1323</td>\n",
       "      <td>ENST00000699524</td>\n",
       "      <td>&gt;ENST00000699524.1 cdna chromosome:GRCh38:1:84...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000122432.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SPATA1</td>\n",
       "      <td>spermatogenesis associated 1 [Source:HGNC Symb...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1</td>\n",
       "      <td>84506386</td>\n",
       "      <td>84567379</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19119</th>\n",
       "      <td>ENST00000699568.1</td>\n",
       "      <td>54</td>\n",
       "      <td>1032</td>\n",
       "      <td>2</td>\n",
       "      <td>55</td>\n",
       "      <td>3</td>\n",
       "      <td>2336</td>\n",
       "      <td>2334</td>\n",
       "      <td>ENST00000699568</td>\n",
       "      <td>&gt;ENST00000699568.1 cdna chromosome:GRCh38:20:1...</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000089091.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>DZANK1</td>\n",
       "      <td>double zinc ribbon and ankyrin repeat domains ...</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>20</td>\n",
       "      <td>18383367</td>\n",
       "      <td>18467030</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19120 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id  five_prime_utr_length  three_prime_utr_length  \\\n",
       "0      ENST00000000233.10                     88                     401   \n",
       "1       ENST00000000412.8                    159                    1457   \n",
       "2      ENST00000000442.11                    225                     777   \n",
       "3       ENST00000001008.6                    170                    2165   \n",
       "4       ENST00000001146.7                     28                    2989   \n",
       "...                   ...                    ...                     ...   \n",
       "19115   ENST00000699088.1                     22                     239   \n",
       "19116   ENST00000699294.1                    487                    2911   \n",
       "19117   ENST00000699333.1                    181                    5137   \n",
       "19118   ENST00000699524.1                    170                    1490   \n",
       "19119   ENST00000699568.1                     54                    1032   \n",
       "\n",
       "       num_five_prime_utr_exons  start_site_pos  cds_start  cds_end  \\\n",
       "0                             1              89          2      544   \n",
       "1                             2             160          3      836   \n",
       "2                             2             226          3     1274   \n",
       "3                             1             171          2     1381   \n",
       "4                             1              29          2     1540   \n",
       "...                         ...             ...        ...      ...   \n",
       "19115                         1              23          2      922   \n",
       "19116                         2             488          3     1676   \n",
       "19117                         1             182          2     1270   \n",
       "19118                         2             171          3     1325   \n",
       "19119                         2              55          3     2336   \n",
       "\n",
       "       cds_length ensembl_transcript_id  \\\n",
       "0             543       ENST00000000233   \n",
       "1             834       ENST00000000412   \n",
       "2            1272       ENST00000000442   \n",
       "3            1380       ENST00000001008   \n",
       "4            1539       ENST00000001146   \n",
       "...           ...                   ...   \n",
       "19115         921       ENST00000699088   \n",
       "19116        1674       ENST00000699294   \n",
       "19117        1269       ENST00000699333   \n",
       "19118        1323       ENST00000699524   \n",
       "19119        2334       ENST00000699568   \n",
       "\n",
       "                                             annotations  ...  \\\n",
       "0      >ENST00000000233.10 cdna chromosome:GRCh38:7:1...  ...   \n",
       "1      >ENST00000000412.8 cdna chromosome:GRCh38:12:8...  ...   \n",
       "2      >ENST00000000442.11 cdna chromosome:GRCh38:11:...  ...   \n",
       "3      >ENST00000001008.6 cdna chromosome:GRCh38:12:2...  ...   \n",
       "4      >ENST00000001146.7 cdna chromosome:GRCh38:2:72...  ...   \n",
       "...                                                  ...  ...   \n",
       "19115  >ENST00000699088.1 cdna chromosome:GRCh38:7:10...  ...   \n",
       "19116  >ENST00000699294.1 cdna chromosome:GRCh38:12:7...  ...   \n",
       "19117  >ENST00000699333.1 cdna chromosome:GRCh38:19:5...  ...   \n",
       "19118  >ENST00000699524.1 cdna chromosome:GRCh38:1:84...  ...   \n",
       "19119  >ENST00000699568.1 cdna chromosome:GRCh38:20:1...  ...   \n",
       "\n",
       "                     gene    gene_biotype transcript_biotype gene_symbol  \\\n",
       "0      ENSG00000004059.11  protein_coding     protein_coding        ARF5   \n",
       "1       ENSG00000003056.8  protein_coding     protein_coding        M6PR   \n",
       "2      ENSG00000173153.17  protein_coding     protein_coding       ESRRA   \n",
       "3       ENSG00000004478.8  protein_coding     protein_coding       FKBP4   \n",
       "4       ENSG00000003137.9  protein_coding     protein_coding     CYP26B1   \n",
       "...                   ...             ...                ...         ...   \n",
       "19115   ENSG00000213413.3  protein_coding     protein_coding       PVRIG   \n",
       "19116  ENSG00000180881.21  protein_coding     protein_coding       CAPS2   \n",
       "19117  ENSG00000251369.10  protein_coding     protein_coding      ZNF550   \n",
       "19118  ENSG00000122432.19  protein_coding     protein_coding      SPATA1   \n",
       "19119  ENSG00000089091.17  protein_coding     protein_coding      DZANK1   \n",
       "\n",
       "                                             description   build chr  \\\n",
       "0      ADP ribosylation factor 5 [Source:HGNC Symbol;...  GRCh38   7   \n",
       "1      mannose-6-phosphate receptor, cation dependent...  GRCh38  12   \n",
       "2      estrogen related receptor alpha [Source:HGNC S...  GRCh38  11   \n",
       "3      FKBP prolyl isomerase 4 [Source:HGNC Symbol;Ac...  GRCh38  12   \n",
       "4      cytochrome P450 family 26 subfamily B member 1...  GRCh38   2   \n",
       "...                                                  ...     ...  ..   \n",
       "19115  PVR related immunoglobulin domain containing [...  GRCh38   7   \n",
       "19116  calcyphosine 2 [Source:HGNC Symbol;Acc:HGNC:16...  GRCh38  12   \n",
       "19117  zinc finger protein 550 [Source:HGNC Symbol;Ac...  GRCh38  19   \n",
       "19118  spermatogenesis associated 1 [Source:HGNC Symb...  GRCh38   1   \n",
       "19119  double zinc ribbon and ankyrin repeat domains ...  GRCh38  20   \n",
       "\n",
       "           start        end  strand  \n",
       "0      127588411  127591700       1  \n",
       "1        8940361    8949645      -1  \n",
       "2       64305524   64316743       1  \n",
       "3        2794970    2805423       1  \n",
       "4       72129238   72147862      -1  \n",
       "...          ...        ...     ...  \n",
       "19115  100219949  100221490       1  \n",
       "19116   75275979   75330324      -1  \n",
       "19117   57541838   57559863      -1  \n",
       "19118   84506386   84567379       1  \n",
       "19119   18383367   18467030      -1  \n",
       "\n",
       "[19120 rows x 22 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/zk/dmbrd2310jn8kdb96r649lr80000gn/T/ipykernel_875/808106407.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  snps['annotation'] = snps['five_prime_UTR_variant_annotation'].apply(parse_five_prime_utr_variant_consequence)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "snps = parsed_variants[parsed_variants['Allele'].isin(['A', 'C', 'G', 'T']) & \n",
    "                       ~parsed_variants['cDNA_position'].str.contains(r'\\d+-\\d+') &\n",
    "                       parsed_variants['Feature'].str.contains('ENST00000358273')\n",
    "                       ]\n",
    "snps['annotation'] = snps['five_prime_UTR_variant_annotation'].apply(parse_five_prime_utr_variant_consequence)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>...</th>\n",
       "      <th>CANONICAL</th>\n",
       "      <th>MANE_SELECT</th>\n",
       "      <th>MANE_PLUS_CLINICAL</th>\n",
       "      <th>existing_InFrame_oORFs</th>\n",
       "      <th>existing_OutOfFrame_oORFs</th>\n",
       "      <th>existing_uORFs</th>\n",
       "      <th>five_prime_UTR_variant_annotation</th>\n",
       "      <th>five_prime_UTR_variant_consequence</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2693</th>\n",
       "      <td>17-31095036-T-G</td>\n",
       "      <td>17:31095036</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000196712</td>\n",
       "      <td>ENST00000358273</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>60</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>YES</td>\n",
       "      <td>NM_001042492.3</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>uSTOP_lost_AltStop:False,uSTOP_lost_AltStopDis...</td>\n",
       "      <td>uSTOP_lost</td>\n",
       "      <td>17-31095036-T-G</td>\n",
       "      <td>{'uSTOP_lost_AltStop': 'False', 'uSTOP_lost_Al...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2694</th>\n",
       "      <td>17-31095036-T-G</td>\n",
       "      <td>17:31095036</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000196712</td>\n",
       "      <td>ENST00000358273</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>60</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>YES</td>\n",
       "      <td>NM_001042492.3</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>uSTOP_lost_AltStop:False,uSTOP_lost_AltStopDis...</td>\n",
       "      <td>uSTOP_lost</td>\n",
       "      <td>17-31095036-T-G</td>\n",
       "      <td>{'uSTOP_lost_AltStop': 'False', 'uSTOP_lost_Al...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2703</th>\n",
       "      <td>17-31095037-A-G</td>\n",
       "      <td>17:31095037</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000196712</td>\n",
       "      <td>ENST00000358273</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>61</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>YES</td>\n",
       "      <td>NM_001042492.3</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>uSTOP_lost_AltStop:False,uSTOP_lost_AltStopDis...</td>\n",
       "      <td>uSTOP_lost</td>\n",
       "      <td>17-31095037-A-G</td>\n",
       "      <td>{'uSTOP_lost_AltStop': 'False', 'uSTOP_lost_Al...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2704</th>\n",
       "      <td>17-31095037-A-G</td>\n",
       "      <td>17:31095037</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000196712</td>\n",
       "      <td>ENST00000358273</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>61</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>YES</td>\n",
       "      <td>NM_001042492.3</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>uSTOP_lost_AltStop:False,uSTOP_lost_AltStopDis...</td>\n",
       "      <td>uSTOP_lost</td>\n",
       "      <td>17-31095037-A-G</td>\n",
       "      <td>{'uSTOP_lost_AltStop': 'False', 'uSTOP_lost_Al...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     #Uploaded_variation     Location Allele             Gene  \\\n",
       "2693     17-31095036-T-G  17:31095036      G  ENSG00000196712   \n",
       "2694     17-31095036-T-G  17:31095036      G  ENSG00000196712   \n",
       "2703     17-31095037-A-G  17:31095037      G  ENSG00000196712   \n",
       "2704     17-31095037-A-G  17:31095037      G  ENSG00000196712   \n",
       "\n",
       "              Feature Feature_type          Consequence cDNA_position  \\\n",
       "2693  ENST00000358273   Transcript  5_prime_UTR_variant            60   \n",
       "2694  ENST00000358273   Transcript  5_prime_UTR_variant            60   \n",
       "2703  ENST00000358273   Transcript  5_prime_UTR_variant            61   \n",
       "2704  ENST00000358273   Transcript  5_prime_UTR_variant            61   \n",
       "\n",
       "     CDS_position Protein_position  ... CANONICAL     MANE_SELECT  \\\n",
       "2693            -                -  ...       YES  NM_001042492.3   \n",
       "2694            -                -  ...       YES  NM_001042492.3   \n",
       "2703            -                -  ...       YES  NM_001042492.3   \n",
       "2704            -                -  ...       YES  NM_001042492.3   \n",
       "\n",
       "     MANE_PLUS_CLINICAL existing_InFrame_oORFs existing_OutOfFrame_oORFs  \\\n",
       "2693                  -                      0                         0   \n",
       "2694                  -                      0                         0   \n",
       "2703                  -                      0                         0   \n",
       "2704                  -                      0                         0   \n",
       "\n",
       "      existing_uORFs                  five_prime_UTR_variant_annotation  \\\n",
       "2693               2  uSTOP_lost_AltStop:False,uSTOP_lost_AltStopDis...   \n",
       "2694               2  uSTOP_lost_AltStop:False,uSTOP_lost_AltStopDis...   \n",
       "2703               2  uSTOP_lost_AltStop:False,uSTOP_lost_AltStopDis...   \n",
       "2704               2  uSTOP_lost_AltStop:False,uSTOP_lost_AltStopDis...   \n",
       "\n",
       "     five_prime_UTR_variant_consequence       variant_id  \\\n",
       "2693                         uSTOP_lost  17-31095036-T-G   \n",
       "2694                         uSTOP_lost  17-31095036-T-G   \n",
       "2703                         uSTOP_lost  17-31095037-A-G   \n",
       "2704                         uSTOP_lost  17-31095037-A-G   \n",
       "\n",
       "                                             annotation  \n",
       "2693  {'uSTOP_lost_AltStop': 'False', 'uSTOP_lost_Al...  \n",
       "2694  {'uSTOP_lost_AltStop': 'False', 'uSTOP_lost_Al...  \n",
       "2703  {'uSTOP_lost_AltStop': 'False', 'uSTOP_lost_Al...  \n",
       "2704  {'uSTOP_lost_AltStop': 'False', 'uSTOP_lost_Al...  \n",
       "\n",
       "[4 rows x 27 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps[\n",
    "    (snps['Allele'] == 'G') &\n",
    "    (snps['five_prime_UTR_variant_consequence'] == 'uSTOP_lost')\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'uSTOP_lost_AltStop:False,uSTOP_lost_AltStopDistanceToCDS:NA,uSTOP_lost_Evidence:NA,uSTOP_lost_FrameWithCDS:outOfFrame,uSTOP_lost_KozakContext:GTGATGG,uSTOP_lost_KozakStrength:Strong'"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps[\n",
    "    (snps['five_prime_UTR_variant_consequence'] == 'uSTOP_lost')\n",
    "]['five_prime_UTR_variant_annotation'].values[15]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "126"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_sequence_and_start(transcript_id):\n",
    "    return sequences[sequences['ensembl_transcript_id'] == 'ENST00000358273']['seq'].values[0], sequences[sequences['ensembl_transcript_id'] == 'ENST00000358273']['start_site_pos'].values[0]\n",
    "\n",
    "seq, start_site = get_sequence_and_start('ENST00000358273')\n",
    "conti_seq = seq[62:]\n",
    "subs = [conti_seq[i:i+3] for i in range(0, len(conti_seq), 3)]\n",
    "subs.index('tga')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'54'"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps[\n",
    "    (snps['Allele'] == 'T') &\n",
    "    (snps['five_prime_UTR_variant_consequence'] == 'uAUG_gained')\n",
    "]['cDNA_position'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "illegal target for annotation (2707551271.py, line 50)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[76], line 50\u001b[0;36m\u001b[0m\n\u001b[0;31m    'annotation-id' : '',\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m illegal target for annotation\n"
     ]
    }
   ],
   "source": [
    "def find_impacted_orf(variant_pos, orfs):\n",
    "    pass\n",
    "\n",
    "\n",
    "def mutate_sequence(cdna_pos, ref, alt, type, seq):\n",
    "    if type == 'SNV':\n",
    "        return seq[:cdna_pos-1] + alt + seq[cdna_pos:]\n",
    "    elif type == 'Deletion':\n",
    "        return seq[:cdna_pos-1] + seq[cdna_pos+len(ref)-1:]\n",
    "    elif type == 'Insertion':\n",
    "        return seq[:cdna_pos-1] + alt + seq[cdna_pos-1:]\n",
    "    else:\n",
    "        raise ValueError('Unknown type')\n",
    "\n",
    "def get_sequence_and_start(transcript_id):\n",
    "    return sequences[sequences['ensembl_transcript_id'] == 'ENST00000358273']['seq'].values[0], sequences[sequences['ensembl_transcript_id'] == 'ENST00000358273']['start_site_pos'].values[0]\n",
    "\n",
    "# uAUG_gained (Deletions, Insertions)\n",
    "# uAUG_lost (SNVs, Deletions, Insertions)\n",
    "# uStop_gained (SNVs, Deletions, Insertions)\n",
    "# uStop_lost (SNVs, Deletions, Insertions)\n",
    "# uFrameshift (SNVs, Deletions, Insertions)\n",
    "\n",
    "def draw_track(consequence, annotation, seq, cdna_pos, ref, alt, start_site, variant_type):\n",
    "    if consequence == 'uAUG_gained':\n",
    "        # Find impacted ORF\n",
    "        new_atg_pos = int(annotation['uAUG_gained_CapDistanceToStart'])\n",
    "        stop_site =  int(annotation['uAUG_gained_CapDistanceToStart']) + int(annotation['uAUG_gained_DistanceToStop'])-1\n",
    "        mutated_seq = mutate_sequence(cdna_pos, ref, alt, variant_type, seq)\n",
    "        new_context = mutated_seq[new_atg_pos-6:new_atg_pos+5]\n",
    "        return {\n",
    "            'annotation_id' : '',\n",
    "            'peturbing_orf' : None,\n",
    "            'context' : {\n",
    "                'ref' : None,\n",
    "                'alt' : new_context,\n",
    "            },\n",
    "            'efficiency':{\n",
    "                'ref_efficiency' : None,\n",
    "                'alt_efficiency' : te[new_context].values[0]\n",
    "            },\n",
    "            'intervals' :{\n",
    "                'start' : new_atg_pos,\n",
    "                'end' : stop_site,\n",
    "            }\n",
    "        }\n",
    "    # UTR Annotator doesn't produce an impact on uAUG lost so we need to get to \n",
    "    # find a way to look at them \n",
    "    elif consequence == 'uAUG_lost':\n",
    "        'annotation-id' : '',\n",
    "        'peturbing_orf' : None,\n",
    "            # Find the orf that is perturbed\n",
    "        'context' : {\n",
    "            'ref' : None,\n",
    "            'alt' : None,\n",
    "        },\n",
    "        'efficiency':{\n",
    "            'ref_efficiency' : None,\n",
    "            'alt_efficiency' : None\n",
    "        },\n",
    "        'intervals' :{\n",
    "            'start' : ,\n",
    "            'end' : ,\n",
    "        }\n",
    "    elif consequence == 'uStop_gained':\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "seq, start_site = get_sequence_and_start('ENST00000358273')\n",
    "mutated = mutate_sequence(115, 'c', 'g', 'SNV', seq)\n",
    "mutated[112-6:112+5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
